src/ray/object_manager/object_manager.cc

```diff
@@ -511,19 +520,27 @@ ray::Status ObjectManager::ExecuteReceiveObject(
     mutable_vec.resize(buffer_length);
     std::vector<boost::asio::mutable_buffer> buffer;
     buffer.push_back(asio::buffer(mutable_vec, buffer_length));
-    boost::system::error_code ec;
-    conn->AsyncReadBuffer(buffer, [](const boost::system::error_code &ec) {
+    conn->AsyncReadBuffer(buffer, [this, conn](const boost::system::error_code &ec) {
       if (ec.value() != 0) {
         RAY_LOG(ERROR) << ec.message();
       }
       // TODO(hme): If the object isn't local, create a pull request for this chunk.
+      conn->ProcessMessages();
+      RAY_CHECK_OK(TransferCompleted(TransferQueue::TransferType::RECEIVE));
     });
   }
-  conn->ProcessMessages();
-  RAY_LOG(DEBUG) << "ReceiveCompleted " << client_id_ << " " << object_id << " "
-                 << num_transfers_receive_ << "/" << config_.max_receives;
-  RAY_CHECK_OK(TransferCompleted(TransferQueue::TransferType::RECEIVE));
   return Status::OK();
 }
 
+flatbuffers::FlatBufferBuilder &ObjectManager::GetFlatBufferBuilder(
+    const UniqueID &fbb_id) {
+  std::unique_lock<std::mutex> guard(flat_buffer_mutex_);
+  return flatbuffer_context_[fbb_id];
```

Comment:
Hmm, is the Flatbuffer builder thread safe? We are returning a reference here, which could be accessed by multiple threads (the guard goes out of scope when the method returns)

Reply:
No two threads access the same FlatBufferBuilder, so this should be okay.

Reply:
Oh ok, so this is to lock the access to the map